src/libteam/0005-libteam-Add-warm_reboot-mode.patch

```diff
@@ -0,0 +1,354 @@
+ 	return err;
+ }
+ 
+-static int lacpdu_recv(struct lacp_port *lacp_port)
++static int lacpdu_process(struct lacp_port *lacp_port, struct lacpdu* lacpdu)
+ {
+-	struct lacpdu lacpdu;
+-	struct sockaddr_ll ll_from;
+ 	int err;
+ 
+-	err = teamd_recvfrom(lacp_port->sock, &lacpdu, sizeof(lacpdu), 0,
+-			     (struct sockaddr *) &ll_from, sizeof(ll_from));
+-	if (err <= 0)
+-		return err;
+-
+-	if (!lacpdu_check(&lacpdu)) {
++	if (!lacpdu_check(lacpdu)) {
+ 		teamd_log_warn("malformed LACP PDU came.");
+ 		return 0;
+ 	}
+ 
++	/* save received lacp pdu frame */
++	(void)memcpy(&lacp_port->last_pdu, lacpdu, sizeof(struct lacpdu));
++	lacp_port->lacpdu_saved = true;
++
+ 	/* Check if we have correct info about the other side */
+-	if (memcmp(&lacpdu.actor, &lacp_port->partner,
++	if (memcmp(&lacpdu->actor, &lacp_port->partner,
+ 		   sizeof(struct lacpdu_info))) {
+-		lacp_port->partner = lacpdu.actor;
++		lacp_port->partner = lacpdu->actor;
+ 		err = lacp_port_partner_update(lacp_port);
+ 		if (err)
+ 			return err;
+@@ -1118,7 +1118,7 @@ static int lacpdu_recv(struct lacp_port *lacp_port)
+ 
+ 	/* Check if the other side has correct info about us */
+ 	if (!lacp_port->periodic_on &&
+-	    memcmp(&lacpdu.partner, &lacp_port->actor,
++	    memcmp(&lacpdu->partner, &lacp_port->actor,
+ 		   sizeof(struct lacpdu_info))) {
+ 		err = lacpdu_send(lacp_port);
+ 		if (err)
+@@ -1133,6 +1133,47 @@ static int lacpdu_recv(struct lacp_port *lacp_port)
+ 	return 0;
+ }
+ 
++static int lacpdu_recv(struct lacp_port *lacp_port)
++{
++	struct lacpdu lacpdu;
++	struct sockaddr_ll ll_from;
++	int err;
++
++	err = teamd_recvfrom(lacp_port->sock, &lacpdu, sizeof(lacpdu), 0,
++			     (struct sockaddr *) &ll_from, sizeof(ll_from));
++	if (err <= 0)
++		return err;
++
++	return lacpdu_process(lacp_port, &lacpdu);
++}
++
++static int lacpdu_read(struct lacp_port *lacp_port)
++{
++	FILE* fp;
++	char path[PATH_MAX];
++	struct lacpdu lacpdu;
++	int err;
++
++	strcpy(path, lacp_port->ctx->lacp_directory);
++	if (path[strlen(path) - 1] != '/')
++		strcat(path, "/"); /* Add trailing slash if we don't have one in the path */
++	strcat(path, lacp_port->tdport->ifname);
++	fp = fopen(path, "r");
++	if (!fp) {
++		teamd_log_err("Can't open lacp-saved dump from file %s: %s", path, strerror(errno));
++		return errno;
++	}
++
++	err = fread(&lacpdu, sizeof(struct lacpdu), 1, fp);
++	(void)fclose(fp);
++	if (err <= 0) {
++		teamd_log_err("Can't read lacp-saved dump from file %s: %s", path, strerror(errno));
++		return err;
++	}
++
++        return lacpdu_process(lacp_port, &lacpdu);
++}
++
+ static int lacp_callback_timeout(struct teamd_context *ctx, int events,
+ 				 void *priv)
+ {
+@@ -1299,6 +1340,11 @@ static int lacp_port_added(struct teamd_context *ctx,
+ 	lacp_port_actor_init(lacp_port);
+ 	lacp_port_link_update(lacp_port);
+ 
++	/* Read data from file and process it */
++	if (ctx->warm_reboot && ctx->lacp_directory) {
++		(void)lacpdu_read(lacp_port);
++	}
```

Comment:
When will the dumped data file for each port be removed?  erase after read or upon port remove, or outside of the teamd process?  Leaving the file permanently there might cause confusion if there are multiple reboot/restart (Mixed warm/cold).

Reply:
it should be outside program to remove the file, teamd already close it and read it.

Reply:
My design is to remove the files right after the read. The information in the file is outdated right after the read.

Reply:
Addressed in the latest patch